xls/modules/rle/rle.x

```diff
@@ -0,0 +1,449 @@
+
+// Simple transaction without overflow
+const CountSymbolTestSymbolWidth = TEST_COMMON_SYMBOL_WIDTH;
+const CountSymbolTestCountWidth  = TEST_COMMON_COUNT_WIDTH;
+
+type CountSymbolTestStimulus   = TestCommonSymbol;
+type CountSymbolTestSymbol     = TestCommonSymbol;
+type CountSymbolTestCount      = TestCommonCount;
+type CountSymbolTestEncInData  = TestCommonEncInData;
+type CountSymbolTestEncOutData = TestCommonEncOutData;
+
+#[test_proc]
+proc RLEEncCountSymbolTest {
+  terminator: chan<bool> out;
+  enc_input_s: chan<CountSymbolTestEncInData> out;
+  enc_output_r: chan<CountSymbolTestEncOutData> in;
+
+  init {()}
+  config (terminator: chan<bool> out) {
+    let (enc_input_s, enc_input_r) = chan<CountSymbolTestEncInData>;
+    let (enc_output_s, enc_output_r) = chan<CountSymbolTestEncOutData>;
+
+    spawn RLEEnc<CountSymbolTestSymbolWidth, CountSymbolTestCountWidth>(
+        enc_input_r, enc_output_s);
+    (terminator, enc_input_s, enc_output_r)
+  }
+  next (tok: token, state:()) {
+    let CountSymbolTestTestStimuli: CountSymbolTestStimulus[4] = [
+      CountSymbolTestStimulus:0xA, CountSymbolTestStimulus:0xA,
+      CountSymbolTestStimulus:0xA, CountSymbolTestStimulus:0xB,
+    ];
+    let tok = for ((counter, symbol), tok):
+                  ((u32, CountSymbolTestStimulus) , token)
+                  in enumerate(CountSymbolTestTestStimuli) {
+      let last = counter == (array_size(CountSymbolTestTestStimuli) - u32:1);
+      let stimulus = CountSymbolTestEncInData{symbol: symbol, last: last};
+      let tok = send(tok, enc_input_s, stimulus);
+      let _ = trace_fmt!("Sent {} stimuli, symbol: 0x{:x}, last: {}",
+      counter, stimulus.symbol, stimulus.last);
+      (tok)
+    }(tok);
+    let CountSymbolTestTestOutput:
+        (CountSymbolTestSymbol, CountSymbolTestCount)[2] = [
+      (CountSymbolTestSymbol:0xA, CountSymbolTestCount:0x3),
+      (CountSymbolTestSymbol:0xB, CountSymbolTestCount:0x1),
+    ];
+    let tok = for ((counter, (symbol, count)), tok):
+        ((u32, (CountSymbolTestSymbol, CountSymbolTestCount)) , token)
+        in enumerate(CountSymbolTestTestOutput) {
+      let last = counter == (array_size(CountSymbolTestTestOutput) - u32:1);
+      let expected = CountSymbolTestEncOutData{
+          symbol: symbol, count: count, last: last};
+      let (tok, enc_output) = recv(tok, enc_output_r);
+      let _ = trace_fmt!(
+        "Received {} pairs, symbol: 0x{:x}, count: {}, last: {}",
+        counter, enc_output.symbol, enc_output.count, enc_output.last
+      );
+      let _ = assert_eq(enc_output, expected);
+      (tok)
+    }(tok);
+    let _ = send(tok, terminator, true);
+    ()
+  }
+}
+
+// Transaction with counter overflow
+const OverflowSymbolWidth = TEST_COMMON_SYMBOL_WIDTH;
+const OverflowCountWidth  = u32:2;
+
+type OverflowStimulus   = TestCommonSymbol;
+type OverflowSymbol     = TestCommonSymbol;
+type OverflowCount      = bits[OverflowCountWidth];
```

Comment:
it would be nice to have support for local `type` in `proc`, should we file a feature request, wdyt?

Reply:
I think it would be great to have `type`s local to `proc`s.
I think it may add new issue with accessing them, maybe `::` syntax, but I don't know.

Reply:
Opened it as #1050